File: src/main/java/brassworks/opac_essentials/client/AddClaimPermissionScreen.java
package brassworks.opac_essentials.client;

import brassworks.opac_essentials.network.ClaimPermissionMutationPayload;
import brassworks.opac_essentials.network.ClaimPermissionsNetwork;
import brassworks.opac_essentials.network.ClaimPermissionsSyncPayload;
import brassworks.opac_essentials.permission.ClaimPermissionAction;
import brassworks.opac_essentials.permission.ClaimPermissionTarget;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.components.EditBox;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import org.lwjgl.glfw.GLFW;

import java.util.Comparator;
import java.util.EnumMap;
import java.util.EnumSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Collectors;

public final class AddClaimPermissionScreen extends Screen {
    private static final int PANEL_HEIGHT = 302;
    private static final int SUGGESTION_HEIGHT = 12;
    private static final int MAX_SUGGESTIONS = 10;

    private final Screen parent;
    private final ClaimPermissionsSyncPayload payload;
    private final EnumSet<ClaimPermissionAction> selectedActions =
            EnumSet.of(ClaimPermissionAction.INTERACT);
    private final EnumMap<ClaimPermissionAction, Button> actionButtons =
            new EnumMap<>(ClaimPermissionAction.class);

    private ClaimPermissionTarget target = ClaimPermissionTarget.BLOCK;
    private EditBox targetIdBox;
    private EditBox playerBox;
    private Button targetButton;
    private Button allActionsButton;
    private List<String> targetSuggestions = List.of();
    private List<String> playerSuggestions = List.of();
    private String validationMessage = "";
    private int panelLeft;
    private int panelTop;
    private int panelWidth;
    private int fieldLeft;
    private int fieldWidth;

    public AddClaimPermissionScreen(
            Screen parent,
            ClaimPermissionsSyncPayload payload
    ) {
        super(Component.translatable(
                "screen.opac_essentials.claim_permissions.add_title"
        ));
        this.parent = parent;
        this.payload = payload;
    }

    @Override
    protected void init() {
        super.init();

        ClaimPermissionTarget.prepareRegisteredIds();

        this.panelWidth = Math.min(340, this.width - 32);
        this.panelLeft = (this.width - this.panelWidth) / 2;
        this.panelTop = Math.max(
                4,
                (this.height - PANEL_HEIGHT) / 2
        );
        this.fieldLeft = this.panelLeft + 14;
        this.fieldWidth = this.panelWidth - 28;

        this.targetButton = this.addRenderableWidget(
                Button.builder(
                                Component.empty(),
                                button -> cycleTarget()
                        )
                        .bounds(
                                this.fieldLeft,
                                this.panelTop + 38,
                                this.fieldWidth,
                                20
                        )
                        .build()
        );

        this.targetIdBox = this.addRenderableWidget(
                new EditBox(
                        this.font,
                        this.fieldLeft,
                        this.panelTop + 82,
                        this.fieldWidth,
                        20,
                        Component.translatable(
                                "screen.opac_essentials.claim_permissions.target_id"
                        )
                )
        );

        this.targetIdBox.setMaxLength(256);
        this.targetIdBox.setHint(Component.literal(""));
        this.targetIdBox.setResponder(
                value -> updateTargetSuggestions()
        );
        this.targetIdBox.setValue("");

        updateTargetSuggestions();

        int actionGap = 4;
        int actionWidth =
                (this.fieldWidth - actionGap * 3) / 4;

        this.allActionsButton = this.addRenderableWidget(
                Button.builder(
                                Component.empty(),
                                button -> toggleAllActions()
                        )
                        .bounds(
                                this.fieldLeft,
                                this.panelTop + 155,
                                actionWidth,
                                20
                        )
                        .build()
        );

        for (ClaimPermissionAction permissionAction
                : ClaimPermissionAction.values()) {
            Button button = this.addRenderableWidget(
                    Button.builder(
                                    Component.empty(),
                                    ignored -> toggleAction(
                                            permissionAction
                                    )
                            )
                            .bounds(
                                    this.fieldLeft,
                                    this.panelTop + 155,
                                    actionWidth,
                                    20
                            )
                            .build()
            );

            this.actionButtons.put(
                    permissionAction,
                    button
            );
        }

        this.playerBox = this.addRenderableWidget(
                new EditBox(
                        this.font,
                        this.fieldLeft,
                        this.panelTop + 198,
                        this.fieldWidth,
                        20,
                        Component.translatable(
                                "screen.opac_essentials.claim_permissions.player"
                        )
                )
        );

        this.playerBox.setMaxLength(128);
        this.playerBox.setResponder(
                value -> updatePlayerSuggestions()
        );
        this.playerBox.setValue("all");

        updatePlayerSuggestions();

        int halfWidth = (this.fieldWidth - 6) / 2;

        this.addRenderableWidget(
                Button.builder(
                                Component.translatable("gui.cancel"),
                                button -> this.onClose()
                        )
                        .bounds(
                                this.fieldLeft,
                                this.panelTop + 264,
                                halfWidth,
                                20
                        )
                        .build()
        );

        this.addRenderableWidget(
                Button.builder(
                                Component.translatable(
                                        "screen.opac_essentials.claim_permissions.add"
                                ),
                                button -> submit()
                        )
                        .bounds(
                                this.fieldLeft + halfWidth + 6,
                                this.panelTop + 264,
                                halfWidth,
                                20
                        )
                        .build()
        );

        updateButtons();
    }

    private void cycleTarget() {
        ClaimPermissionTarget[] targets =
                ClaimPermissionTarget.values();

        this.target = targets[
                (this.target.ordinal() + 1) % targets.length
                ];

        List<ClaimPermissionAction> supported =
                ClaimPermissionsScreen.supportedActions(
                        this.target
                );

        this.selectedActions.retainAll(supported);

        if (this.selectedActions.isEmpty()) {
            this.selectedActions.add(
                    supported.getFirst()
            );
        }

        ResourceLocation current =
                ResourceLocation.tryParse(
                        this.targetIdBox
                                .getValue()
                                .trim()
                );

        if (current == null
                || !this.target.isRegistered(current)) {
            this.targetIdBox.setValue("");
        }

        updateTargetSuggestions();
        updateButtons();
    }

    private void toggleAction(
            ClaimPermissionAction permissionAction
    ) {
        if (!this.target.supports(permissionAction)) {
            return;
        }

        if (!this.selectedActions.remove(
                permissionAction
        )) {
            this.selectedActions.add(
                    permissionAction
            );
        }

        updateButtons();
    }

    private void toggleAllActions() {
        List<ClaimPermissionAction> supported =
                ClaimPermissionsScreen.supportedActions(
                        this.target
                );

        if (this.selectedActions.containsAll(
                supported
        )) {
            this.selectedActions.removeAll(
                    supported
            );
        } else {
            this.selectedActions.addAll(
                    supported
            );
        }

        updateButtons();
    }

    private void updateButtons() {
        this.targetButton.setMessage(
                Component.literal(
                        ClaimPermissionsScreen.displayName(
                                this.target.name()
                        )
                )
        );

        List<ClaimPermissionAction> supported =
                ClaimPermissionsScreen.supportedActions(
                        this.target
                );

        boolean allSelected =
                this.selectedActions.containsAll(
                        supported
                );

        this.allActionsButton.setMessage(
                Component.literal(
                        selectionPrefix(allSelected) + "All"
                )
        );

        int actionGap = 4;
        int actionWidth =
                (this.fieldWidth - actionGap * 3) / 4;
        int visibleIndex = 1;

        for (ClaimPermissionAction permissionAction
                : ClaimPermissionAction.values()) {
            Button button =
                    this.actionButtons.get(
                            permissionAction
                    );

            button.visible =
                    supported.contains(
                            permissionAction
                    );

            if (!button.visible) {
                continue;
            }

            button.setX(
                    this.fieldLeft
                            + visibleIndex
                            * (actionWidth + actionGap)
            );

            button.setMessage(
                    Component.literal(
                            selectionPrefix(
                                    this.selectedActions.contains(
                                            permissionAction
                                    )
                            )
                                    + ClaimPermissionsScreen.displayName(
                                    permissionAction.name()
                            )
                    )
            );

            visibleIndex++;
        }
    }

    private static String selectionPrefix(
            boolean selected
    ) {
        return selected ? "[x] " : "[ ] ";
    }

    private void updateTargetSuggestions() {
        String query = this.targetIdBox
                .getValue()
                .trim()
                .toLowerCase(Locale.ROOT);

        this.targetSuggestions =
                this.target.registeredIds()
                        .map(ResourceLocation::toString)
                        .filter(id ->
                                matchesRegistryId(
                                        id,
                                        query
                                )
                        )
                        .filter(id ->
                                !id.equalsIgnoreCase(
                                        query
                                )
                        )
                        .sorted(
                                Comparator
                                        .comparingInt(
                                                (String id) ->
                                                        matchRank(
                                                                id,
                                                                query
                                                        )
                                        )
                                        .thenComparing(
                                                Comparator.naturalOrder()
                                        )
                        )
                        .limit(MAX_SUGGESTIONS)
                        .toList();

        updateInlineSuggestion(
                this.targetIdBox,
                this.targetSuggestions
        );
    }

    private static boolean matchesRegistryId(
            String id,
            String query
    ) {
        if (query.isEmpty()) {
            return true;
        }

        int separator = id.indexOf(':');
        String path = separator < 0
                ? id
                : id.substring(separator + 1);

        return id.startsWith(query)
                || path.startsWith(query)
                || id.contains(query);
    }

    private static int matchRank(
            String id,
            String query
    ) {
        if (id.startsWith(query)) {
            return 0;
        }

        int separator = id.indexOf(':');
        String path = separator < 0
                ? id
                : id.substring(separator + 1);

        return path.startsWith(query)
                ? 1
                : 2;
    }

    private void updatePlayerSuggestions() {
        Set<String> playerNames =
                new TreeSet<>(
                        String.CASE_INSENSITIVE_ORDER
                );

        playerNames.add("all");

        if (this.minecraft.getConnection() != null) {
            this.minecraft
                    .getConnection()
                    .getOnlinePlayers()
                    .forEach(
                            playerInfo ->
                                    playerNames.add(
                                            playerInfo
                                                    .getProfile()
                                                    .getName()
                                    )
                    );
        }

        String query = this.playerBox
                .getValue()
                .trim()
                .toLowerCase(Locale.ROOT);

        this.playerSuggestions =
                playerNames.stream()
                        .filter(name ->
                                name.toLowerCase(
                                                Locale.ROOT
                                        )
                                        .startsWith(query)
                        )
                        .filter(name ->
                                !name.equalsIgnoreCase(
                                        query
                                )
                        )
                        .limit(MAX_SUGGESTIONS)
                        .toList();

        updateInlineSuggestion(
                this.playerBox,
                this.playerSuggestions
        );
    }

    private static void updateInlineSuggestion(
            EditBox box,
            List<String> suggestions
    ) {
        String value = box.getValue();

        if (!suggestions.isEmpty()
                && suggestions
                .getFirst()
                .regionMatches(
                        true,
                        0,
                        value,
                        0,
                        value.length()
                )) {
            box.setSuggestion(
                    suggestions
                            .getFirst()
                            .substring(
                                    value.length()
                            )
            );
        } else {
            box.setSuggestion(null);
        }
    }

    private void submit() {
        ResourceLocation targetId =
                ResourceLocation.tryParse(
                        this.targetIdBox
                                .getValue()
                                .trim()
                );

        if (targetId == null) {
            this.validationMessage =
                    "Invalid registry ID.";
            return;
        }

        if (!this.target.isRegistered(targetId)) {
            this.validationMessage =
                    "That registry ID does not exist.";
            return;
        }

        List<ClaimPermissionAction> actions =
                ClaimPermissionsScreen
                        .supportedActions(this.target)
                        .stream()
                        .filter(
                                this.selectedActions::contains
                        )
                        .toList();

        if (actions.isEmpty()) {
            this.validationMessage =
                    "Select at least one action.";
            return;
        }

        String playerName =
                this.playerBox
                        .getValue()
                        .trim();

        if (playerName.isEmpty()) {
            playerName = "all";
        }

        String actionNames =
                actions.stream()
                        .map(Enum::name)
                        .collect(
                                Collectors.joining(",")
                        );

        ClaimPermissionsNetwork.sendToServer(
                new ClaimPermissionMutationPayload(
                        ClaimPermissionMutationPayload.Operation.ADD,
                        this.payload.claimOwner(),
                        this.payload.subConfigIndex(),
                        this.target.name(),
                        targetId.toString(),
                        actionNames,
                        "",
                        "",
                        playerName
                )
        );

        this.validationMessage =
                actions.size() == 1
                        ? "Saving permission..."
                        : "Saving "
                          + actions.size()
                          + " permissions...";
    }

    @Override
    public boolean keyPressed(
            int keyCode,
            int scanCode,
            int modifiers
    ) {
        if (keyCode == GLFW.GLFW_KEY_TAB) {
            if (acceptFirstSuggestion(
                    this.targetIdBox,
                    this.targetSuggestions
            )) {
                return true;
            }

            if (acceptFirstSuggestion(
                    this.playerBox,
                    this.playerSuggestions
            )) {
                return true;
            }
        }

        return super.keyPressed(
                keyCode,
                scanCode,
                modifiers
        );
    }

    private static boolean acceptFirstSuggestion(
            EditBox box,
            List<String> suggestions
    ) {
        if (box.isFocused()
                && !suggestions.isEmpty()) {
            box.setValue(
                    suggestions.getFirst()
            );
            return true;
        }

        return false;
    }

    @Override
    public boolean mouseClicked(
            double mouseX,
            double mouseY,
            int button
    ) {
        if (button == 0) {
            if (acceptClickedSuggestion(
                    this.targetIdBox,
                    this.targetSuggestions,
                    this.panelTop + 104,
                    mouseX,
                    mouseY
            )) {
                return true;
            }

            if (acceptClickedSuggestion(
                    this.playerBox,
                    this.playerSuggestions,
                    this.panelTop + 220,
                    mouseX,
                    mouseY
            )) {
                return true;
            }
        }

        return super.mouseClicked(
                mouseX,
                mouseY,
                button
        );
    }

    private boolean acceptClickedSuggestion(
            EditBox box,
            List<String> suggestions,
            int top,
            double mouseX,
            double mouseY
    ) {
        if (!box.isFocused()
                || mouseX < this.fieldLeft
                || mouseX >= this.fieldLeft
                + this.fieldWidth
                || mouseY < top) {
            return false;
        }

        int index = (int) (
                (mouseY - top)
                        / SUGGESTION_HEIGHT
        );

        if (index < 0
                || index >= suggestions.size()) {
            return false;
        }

        box.setValue(
                suggestions.get(index)
        );
        box.setFocused(true);

        return true;
    }

    @Override
    public void onClose() {
        this.minecraft.setScreen(
                this.parent
        );
    }

    @Override
    public void render(
            GuiGraphics graphics,
            int mouseX,
            int mouseY,
            float partialTick
    ) {
        super.render(
                graphics,
                mouseX,
                mouseY,
                partialTick
        );

        if (this.targetIdBox.isFocused()) {
            renderSuggestions(
                    graphics,
                    this.targetSuggestions,
                    this.panelTop + 104,
                    mouseX,
                    mouseY
            );
        }

        if (this.playerBox.isFocused()) {
            renderSuggestions(
                    graphics,
                    this.playerSuggestions,
                    this.panelTop + 220,
                    mouseX,
                    mouseY
            );
        }
    }

    @Override
    public void renderBackground(
            GuiGraphics graphics,
            int mouseX,
            int mouseY,
            float partialTick
    ) {
        super.renderBackground(
                graphics,
                mouseX,
                mouseY,
                partialTick
        );

        graphics.fill(
                this.panelLeft,
                this.panelTop,
                this.panelLeft + this.panelWidth,
                this.panelTop + PANEL_HEIGHT,
                0xE0101010
        );

        graphics.drawCenteredString(
                this.font,
                this.title,
                this.width / 2,
                this.panelTop + 10,
                0xFFFFFF
        );

        drawLabel(
                graphics,
                "screen.opac_essentials.claim_permissions.category",
                27
        );

        drawLabel(
                graphics,
                "screen.opac_essentials.claim_permissions.target_id",
                70
        );

        drawLabel(
                graphics,
                "screen.opac_essentials.claim_permissions.action",
                143
        );

        drawLabel(
                graphics,
                "screen.opac_essentials.claim_permissions.player",
                186
        );

        if (!this.validationMessage.isBlank()) {
            graphics.drawCenteredString(
                    this.font,
                    this.validationMessage,
                    this.width / 2,
                    this.panelTop + 289,
                    0xFFAA55
            );
        }
    }

    private void drawLabel(
            GuiGraphics graphics,
            String translationKey,
            int yOffset
    ) {
        graphics.drawString(
                this.font,
                Component.translatable(
                        translationKey
                ),
                this.fieldLeft,
                this.panelTop + yOffset,
                0xA0A0A0,
                false
        );
    }

    private void renderSuggestions(
            GuiGraphics graphics,
            List<String> suggestions,
            int top,
            int mouseX,
            int mouseY
    ) {
        for (int index = 0;
             index < suggestions.size();
             index++) {
            int y = top
                    + index
                    * SUGGESTION_HEIGHT;

            boolean hovered =
                    mouseX >= this.fieldLeft
                            && mouseX
                            < this.fieldLeft
                            + this.fieldWidth
                            && mouseY >= y
                            && mouseY
                            < y
                            + SUGGESTION_HEIGHT;

            graphics.fill(
                    this.fieldLeft,
                    y,
                    this.fieldLeft + this.fieldWidth,
                    y + SUGGESTION_HEIGHT,
                    hovered
                            ? 0xF0404040
                            : 0xF0181818
            );

            String suggestion =
                    this.font.plainSubstrByWidth(
                            suggestions.get(index),
                            this.fieldWidth - 8
                    );

            graphics.drawString(
                    this.font,
                    suggestion,
                    this.fieldLeft + 4,
                    y + 2,
                    0xFFFFFF,
                    false
            );
        }
    }
}

File: src/main/java/brassworks/opac_essentials/claims/permission/client/screen/AddClaimPermissionScreen.kt
package brassworks.opac_essentials.claims.permission.client.screen

import brassworks.opac_essentials.claims.permission.network.ClaimPermissionMutationPayload
import brassworks.opac_essentials.claims.permission.network.ClaimPermissionsNetwork
import brassworks.opac_essentials.claims.permission.network.ClaimPermissionsSyncPayload
import brassworks.opac_essentials.claims.permission.model.ClaimPermissionAction
import brassworks.opac_essentials.claims.permission.model.ClaimPermissionTarget
import gg.essential.elementa.components.UIContainer
import gg.essential.elementa.constraints.CenterConstraint
import gg.essential.elementa.dsl.basicHeightConstraint
import gg.essential.elementa.dsl.basicWidthConstraint
import gg.essential.elementa.dsl.childOf
import gg.essential.elementa.dsl.constrain
import gg.essential.elementa.dsl.minus
import gg.essential.elementa.dsl.percent
import gg.essential.elementa.dsl.pixels
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.screens.Screen
import net.minecraft.resources.ResourceLocation
import net.swzo.brass.ui.BrassScreen
import net.swzo.brass.ui.BrassThemes
import net.swzo.brass.ui.Colors
import net.swzo.brass.ui.kit.base.BrassAccent
import net.swzo.brass.ui.kit.input.BrassButton
import net.swzo.brass.ui.kit.input.BrassCheckbox
import net.swzo.brass.ui.kit.surface.BrassPanel
import net.swzo.brass.ui.kit.surface.BrassWindow
import net.swzo.brass.ui.kit.text.BrassLabel
import net.swzo.brass.ui.kit.text.BrassTextInput
import java.awt.Color
import java.util.EnumSet
import java.util.Locale

class AddClaimPermissionScreen(
    private val parent: Screen,
    private val payload: ClaimPermissionsSyncPayload,
) : BrassScreen(backdropColor = Color(22, 14, 11, 148)) {

    private var target = ClaimPermissionTarget.BLOCK
    private val selectedActions = EnumSet.of(ClaimPermissionAction.INTERACT)
    private val actionCheckboxes = mutableMapOf<ClaimPermissionAction, BrassCheckbox>()
    private var saving = false

    private lateinit var categoryOptions: UIContainer
    private lateinit var targetInput: BrassTextInput
    private lateinit var targetSuggestions: UIContainer
    private lateinit var actionOptions: UIContainer
    private lateinit var playerInput: BrassTextInput
    private lateinit var playerSuggestions: UIContainer
    private lateinit var validationLabel: BrassLabel
    private lateinit var submitButton: BrassButton

    init {
        BrassThemes.apply("mocha")
        ClaimPermissionTarget.prepareRegisteredIds()
        buildUi()
        rebuildCategories()
        rebuildActions()
        refreshTargetSuggestions()
        refreshPlayerSuggestions()
        refreshValidation()
    }

    private fun buildUi() {
        val frame = BrassWindow(
            title = "OPAC / Claims",
            subtitle = "Add permission",
            onClose = ::returnToParent,
            controls = false,
            minW = 340f,
            minH = 330f,
        ).constrain {
            x = CenterConstraint()
            y = CenterConstraint()
            width = basicWidthConstraint { component ->
                minOf(WINDOW_WIDTH, (component.parent.getWidth() - 32f).coerceAtLeast(320f))
            }
            height = basicHeightConstraint { component ->
                minOf(WINDOW_HEIGHT, (component.parent.getHeight() - 24f).coerceAtLeast(310f))
            }
        } childOf background

        BrassLabel("Add permissions to ${payload.scopeName()}", Colors.UI_TEXT_DARK).constrain {
            x = 12.pixels()
            y = 8.pixels()
        } childOf frame.content

        val formPanel = BrassPanel(
            title = "NEW PERMISSION",
            layout = BrassPanel.Layout.FREE,
        ).constrain {
            x = 12.pixels()
            y = 27.pixels()
            width = 100.percent() - 24.pixels()
            height = 100.percent() - 63.pixels()
        } childOf frame.content

        BrassLabel("CATEGORY", Colors.UI_TEXT_DARK).constrain {
            y = 0.pixels()
        } childOf formPanel.content

        categoryOptions = UIContainer().constrain {
            y = 13.pixels()
            width = 100.percent()
            height = 20.pixels()
        } childOf formPanel.content

        BrassLabel("TARGET ID", Colors.UI_TEXT_DARK).constrain {
            y = 42.pixels()
        } childOf formPanel.content

        targetInput = BrassTextInput(
            initial = "",
            placeholder = "minecraft:acacia_button",
        ) { value ->
            refreshTargetSuggestions(value)
            refreshValidation()
        }.constrain {
            y = 55.pixels()
            width = 100.percent()
            height = 19.pixels()
        } childOf formPanel.content

        targetSuggestions = UIContainer().constrain {
            y = 78.pixels()
            width = 100.percent()
            height = 43.pixels()
        } childOf formPanel.content

        BrassLabel("ACTIONS", Colors.UI_TEXT_DARK).constrain {
            y = 126.pixels()
        } childOf formPanel.content

        actionOptions = UIContainer().constrain {
            y = 140.pixels()
            width = 100.percent()
            height = 22.pixels()
        } childOf formPanel.content

        BrassLabel("PLAYER", Colors.UI_TEXT_DARK).constrain {
            y = 170.pixels()
        } childOf formPanel.content

        playerInput = BrassTextInput(
            initial = "all",
            placeholder = "all or player name",
        ) { value ->
            refreshPlayerSuggestions(value)
            refreshValidation()
        }.constrain {
            y = 183.pixels()
            width = 100.percent()
            height = 19.pixels()
        } childOf formPanel.content

        playerSuggestions = UIContainer().constrain {
            y = 206.pixels()
            width = 100.percent()
            height = 31.pixels()
        } childOf formPanel.content

        validationLabel = BrassLabel("", Colors.UI_TEXT_DARK, scale = 0.9f).constrain {
            y = 100.percent() - 14.pixels()
        } childOf formPanel.content

        BrassButton("Cancel", BrassAccent.DEFAULT) { returnToParent() }.constrain {
            x = 12.pixels()
            y = 100.percent() - 27.pixels()
            width = 72.pixels()
            height = 18.pixels()
        } childOf frame.content

        submitButton = BrassButton("Add permission", BrassAccent.BRASS) { submit() }.constrain {
            x = 100.percent() - 106.pixels()
            y = 100.percent() - 27.pixels()
            width = 94.pixels()
            height = 18.pixels()
        } childOf frame.content
    }

    private fun rebuildCategories() {
        categoryOptions.clearChildren()
        ClaimPermissionTarget.entries.forEachIndexed { index, option ->
            val selected = option == target
            BrassButton(
                label = ClaimPermissionsScreen.displayName(option.name),
                accent = if (selected) BrassAccent.BRASS else BrassAccent.DEFAULT,
            ) {
                selectTarget(option)
            }.also {
                it.selectable = true
                it.selected = selected
            }.constrain {
                x = (index * 25).percent()
                width = 25.percent() - 3.pixels()
                height = 20.pixels()
            } childOf categoryOptions
        }
    }

    private fun selectTarget(option: ClaimPermissionTarget) {
        if (option == target) return
        target = option

        val supported = ClaimPermissionsScreen.supportedActions(target)
        selectedActions.retainAll(supported.toSet())
        if (selectedActions.isEmpty() && supported.isNotEmpty()) {
            selectedActions.add(supported.first())
        }

        val currentId = ResourceLocation.tryParse(targetInput.text.trim())
        if (currentId == null || !target.isRegistered(currentId)) {
            targetInput.value = ""
        }

        rebuildCategories()
        rebuildActions()
        refreshTargetSuggestions()
        refreshValidation()
    }

    private fun rebuildActions() {
        actionOptions.clearChildren()
        actionCheckboxes.clear()
        val supported = ClaimPermissionsScreen.supportedActions(target)

        supported.forEachIndexed { index, action ->
            val row = UIContainer().constrain {
                x = (index * 100f / supported.size).percent()
                width = (100f / supported.size).percent()
                height = 22.pixels()
            } childOf actionOptions

            val checkbox = BrassCheckbox(initial = selectedActions.contains(action)) { checked ->
                if (checked) selectedActions.add(action) else selectedActions.remove(action)
                refreshValidation()
            }.constrain {
                x = 1.pixels()
                y = CenterConstraint()
                width = 14.pixels()
                height = 14.pixels()
            } childOf row
            actionCheckboxes[action] = checkbox

            BrassLabel(ClaimPermissionsScreen.displayName(action.name), Colors.UI_TEXT).constrain {
                x = 20.pixels()
                y = CenterConstraint()
            } childOf row
        }
    }

    private fun refreshTargetSuggestions(value: String = targetInput.text) {
        targetSuggestions.clearChildren()
        val query = value.trim().lowercase(Locale.ROOT)
        if (query.isBlank()) return

        target.registeredIds()
            .map(ResourceLocation::toString)
            .filter { id -> matchesRegistryId(id, query) && !id.equals(query, ignoreCase = true) }
            .sorted(compareBy<String> { matchRank(it, query) }.thenBy { it })
            .limit(MAX_TARGET_SUGGESTIONS.toLong())
            .toList()
            .forEachIndexed { index, id ->
                BrassButton(id) {
                    targetInput.value = id
                    targetSuggestions.clearChildren()
                    refreshValidation()
                }.constrain {
                    y = (index * 14).pixels()
                    width = 100.percent()
                    height = 13.pixels()
                } childOf targetSuggestions
            }
    }

    private fun refreshPlayerSuggestions(value: String = playerInput.text) {
        playerSuggestions.clearChildren()
        val query = value.trim().lowercase(Locale.ROOT)
        val names = sortedSetOf(String.CASE_INSENSITIVE_ORDER, "all")
        Minecraft.getInstance().connection?.onlinePlayers?.forEach { player ->
            names.add(player.profile.name)
        }

        names.asSequence()
            .filter { it.lowercase(Locale.ROOT).startsWith(query) }
            .filter { !it.equals(query, ignoreCase = true) }
            .take(MAX_PLAYER_SUGGESTIONS)
            .forEachIndexed { index, name ->
                BrassButton(name) {
                    playerInput.value = name
                    playerSuggestions.clearChildren()
                    refreshValidation()
                }.constrain {
                    y = (index * 14).pixels()
                    width = 100.percent()
                    height = 13.pixels()
                } childOf playerSuggestions
            }
    }

    private fun refreshValidation() {
        if (!::validationLabel.isInitialized || !::submitButton.isInitialized) return

        val message = validationProblem()
        if (message != null) {
            validationLabel.text = message
            validationLabel.tint = if (targetInput.text.isBlank()) Colors.UI_TEXT_DARK else Colors.DANGER
            submitButton.active = false
            return
        }

        val player = playerInput.text.trim().ifBlank { "all" }
        val count = selectedActions.count { target.supports(it) }
        validationLabel.text = "Ready: $count ${if (count == 1) "permission" else "permissions"} for $player"
        validationLabel.tint = Colors.UI_TEXT_DARK
        submitButton.active = !saving
    }

    private fun validationProblem(): String? {
        val rawId = targetInput.text.trim()
        if (rawId.isBlank()) return "Choose or enter a target ID."

        val targetId = ResourceLocation.tryParse(rawId)
            ?: return "Invalid registry ID."
        if (!target.isRegistered(targetId)) return "That registry ID does not exist in this category."
        if (selectedActions.none(target::supports)) return "Select at least one action."
        return null
    }

    private fun submit() {
        if (saving || validationProblem() != null) {
            refreshValidation()
            return
        }

        val targetId = ResourceLocation.tryParse(targetInput.text.trim()) ?: return
        val actions = ClaimPermissionsScreen.supportedActions(target)
            .filter(selectedActions::contains)
        val playerName = playerInput.text.trim().ifBlank { "all" }

        saving = true
        actionCheckboxes.values.forEach { it.active = false }
        submitButton.active = false
        validationLabel.text = if (actions.size == 1) "Saving permission…" else "Saving ${actions.size} permissions…"
        validationLabel.tint = Colors.UI_TEXT_DARK

        ClaimPermissionsNetwork.sendToServer(
            ClaimPermissionMutationPayload(
                ClaimPermissionMutationPayload.Operation.ADD,
                payload.claimOwner(),
                payload.subConfigIndex(),
                target.name,
                targetId.toString(),
                actions.joinToString(",") { it.name },
                "",
                "",
                playerName,
            ),
        )
    }

    private fun returnToParent() {
        Minecraft.getInstance().setScreen(parent)
    }

    private fun matchesRegistryId(id: String, query: String): Boolean {
        val path = id.substringAfter(':', id)
        return id.startsWith(query) || path.startsWith(query) || id.contains(query)
    }

    private fun matchRank(id: String, query: String): Int = when {
        id.startsWith(query) -> 0
        id.substringAfter(':', id).startsWith(query) -> 1
        else -> 2
    }

    companion object {
        private const val WINDOW_WIDTH = 430f
        private const val WINDOW_HEIGHT = 370f
        private const val MAX_TARGET_SUGGESTIONS = 3
        private const val MAX_PLAYER_SUGGESTIONS = 2
    }
}
